### PR TITLE
fix: improved DataDiscoveryCLI interface

### PR DESCRIPTION
**Pull request by valsdav**
Improving the DataDiscoveryCLI

- `do_replicas` and `load_dataset_definitions` now returns uproot compatible metadata
- Added "first" option for rucio replicas
- Returning both overall and updated datasets from preprocess
- Fixes and docs

@iasonkrom

### CODE REVIEW

File: src/coffea/dataset_tools/dataset_query.py [hunk 1]
```diff
@@ -553,7 +571,11 @@ def load_dataset_definition(
         and selected results and replicas following the options.
 
         - query_results_strategy:  "all" or "manual" to be prompt for selection
-        - replicas_strategy: "round-robin", "choose" (to manually choose the sites), "manual": to be prompt for manual decision case by case
+        - replicas_strategy:
+            - "round-robin": select randomly from the available sites for each file
+            - "choose": filter the sites with a list of indices for all the files
+            - "first": take the first risult returned by rucio
```
Comment (ikrommyd) [inline src/coffea/dataset_tools/dataset_query.py]:
```suggestion
            - "first": take the first result returned by rucio
```

Comment (ikrommyd) [reply src/coffea/dataset_tools/dataset_query.py]: Just saw this typo while looking at the changes, will give it a spin later. Thanks @valsdav!

File: src/coffea/dataset_tools/dataset_query.py [hunk 2]
```diff
@@ -578,16 +600,17 @@ def load_dataset_definition(
...
         else:
-            if replicas_strategy not in ["round-robin", "choose"]:
+            if replicas_strategy not in ["round-robin", "choose", "first"]:
                 print(
-                    "Invalid replicas-strategy: please choose between manual|round-robin|choose"
+                    "Invalid replicas-strategy: please choose between manual|round-robin|choose|first"
                 )
                 exit(1)
-            self.do_replicas(mode=replicas_strategy, selection="all")
+            out_replicas = self.do_replicas(mode=replicas_strategy, selection="all")
```
Comment (ikrommyd) [inline src/coffea/dataset_tools/dataset_query.py:608]: If I'm seeing correctly, `do_replicas()` doesn't return anything so `out_replicas` will be `None`

**Comment by ikrommyd** [comment]
Should we have "first" or "round-robin" be the default though? Opinions?

**Comment by valsdav** [comment]
`round-robin` by default distributes the load more between sites.

**Comment by ikrommyd** [comment]
Does it "reduce" the quality of the sites though? Are "bigger" sites generally better?

**Comment by lgray** [comment]
Round-robin is fine and probably scales better. If it turns out that it causes problems we can change it later, but I think that gets determined by experimentation rather than how we think it is supposed to work.

**Comment by ikrommyd** [comment]
One final thing to ask is a method that returns the dictionary that `do_save` would write to a json file. Maybe `.as_dict` like @lgray mentioned in Slack. After that, it looks good to me!

**Comment by lgray** [comment]
Either `as_dict` or `to_dict` would be fine, but yes, that's needed @valsdav .

**Comment by ikrommyd** [comment]
Oh I got one more thought. What if we also keep `out_available` and `out_updated` dictionaries as class attributes if preprocess was run? Just to have direct access to them from the class and not only from the json files.

File: src/coffea/dataset_tools/dataset_query.py [hunk 3]
```diff
@@ -542,19 +544,19 @@ def do_preprocess(
             "[red] Preprocessing files to extract available chunks with dask[/]"
         ):
             with Client(dask_cluster) as _:
-                out_available, out_updated = preprocess(
+                self.preprocess_available, self.preprocessed_total = preprocess(
```
Comment (ikrommyd) [inline src/coffea/dataset_tools/dataset_query.py]: `self.preprocessed_available` not `preprocess_available`

File: src/coffea/dataset_tools/dataset_query.py [hunk 4]
```diff
@@ -542,19 +544,19 @@ def do_preprocess(
...
                 )
         with gzip.open(f"{output_file}_available.json.gz", "wt") as file:
             print(f"Saved available fileset chunks to {output_file}_available.json.gz")
-            json.dump(out_available, file, indent=2)
+            json.dump(self.preprocessed_total, file, indent=2)
         with gzip.open(f"{output_file}_all.json.gz", "wt") as file:
             print(f"Saved all fileset chunks to {output_file}_all.json.gz")
-            json.dump(out_updated, file, indent=2)
-        return out_available, out_updated
+            json.dump(self.preprocess_available, file, indent=2)
```
Comment (ikrommyd) [inline src/coffea/dataset_tools/dataset_query.py]: same here

File: src/coffea/dataset_tools/dataset_query.py [hunk 5]
```diff
@@ -542,19 +544,19 @@ def do_preprocess(
...
+        return self.preprocessed_total, self.preprocess_available
```
Comment (ikrommyd) [inline src/coffea/dataset_tools/dataset_query.py]: same here

**Comment by lgray** [comment]
@valsdav @iasonkrom ping me when you've converged, and I'll merge it.

**Comment by ikrommyd** [comment]
@lgray `do_preprocess()` takes in `dask_cluster=None` and spawns a `with Client(dask_cluster) as _` to do the preprocessing. But what if the user already has a `Client` running? No reason to spawn another one right? I'm good with all the rest

**Comment by ikrommyd** [comment]
I don't know if passing in `dask_client` and not `dask_cluster` is the correct approach here

**Comment by valsdav** [comment]
if you provide a dask scheduler url the Client is going to connect to that. if None it will spawn a local multiprocessing cluster, isn't it?

**Comment by ikrommyd** [comment]
I'm not sure about that. Lindsey will have to confirm that it's fine. I'm good with all the rest. @lgray you're free to review

**Comment by lgray** [comment]
That's all the functionality we should expect in a CLI system, see https://distributed.dask.org/en/stable/client.html.

So I think we can provide the scheduler endpoint as a string and call it a day here.

If you need to do more complicated preprocessing, you can get the raw list and do what you want, or call the pre-processing command in a script in the appropriate context manager with your dask cluster setup.

What's there is fit to task.

**Comment by lgray** [comment]
@valsdav change `dask_cluster` to `scheduler_url` and I think the usage becomes more clear.

**Comment by ikrommyd** [comment]
so if I've spawned an lpc condor client with:
```python
from distributed import Client
from lpcjobqueue import LPCCondorCluster


cluster = LPCCondorCluster()
cluster.adapt(minimum=0, maximum=100)
client = Client(cluster)
```
what should I do to make `do_preprocess` use that `Client`?

**Comment by lgray** [comment]
`as_dict` -> call `coffea.dataset_tools.preprocess`

**Comment by ikrommyd** [comment]
> `as_dict` -> call `coffea.dataset_tools.preprocess`

ah okay, so we just don't use the CLI class's method in that case. Thanks

I'm good with merging as soon as you review! Thanks a lot @valsdav!

**Pull request merged**